File: ActivityTracker/LogViewController.m
/**
 * LogViewController.m
 * ActivityTracker
 *
 * Created by Yu Suo on 11/5/14.
 * Copyright 2014-2015 MbientLab Inc. All rights reserved.
 *
 * IMPORTANT: Your use of this Software is limited to those specific rights
 * granted under the terms of a software license agreement between the user who
 * downloaded the software, his/her employer (which must be your employer) and
 * MbientLab Inc, (the "License").  You may not use this Software unless you
 * agree to abide by the terms of the License which can be found at
 * www.mbientlab.com/terms.  The License limits your use, and you acknowledge,
 * that the Software may be modified, copied, and distributed when used in
 * conjunction with an MbientLab Inc, product.  Other than for the foregoing
 * purpose, you may not use, reproduce, copy, prepare derivative works of,
 * modify, distribute, perform, display or sell this Software and/or its
 * documentation for any purpose.
 *
 * YOU FURTHER ACKNOWLEDGE AND AGREE THAT THE SOFTWARE AND DOCUMENTATION ARE
 * PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND, EITHER EXPRESS OR IMPLIED,
 * INCLUDING WITHOUT LIMITATION, ANY WARRANTY OF MERCHANTABILITY, TITLE,
 * NON-INFRINGEMENT AND FITNESS FOR A PARTICULAR PURPOSE. IN NO EVENT SHALL
 * MBIENTLAB OR ITS LICENSORS BE LIABLE OR OBLIGATED UNDER CONTRACT, NEGLIGENCE,
 * STRICT LIABILITY, CONTRIBUTION, BREACH OF WARRANTY, OR OTHER LEGAL EQUITABLE
 * THEORY ANY DIRECT OR INDIRECT DAMAGES OR EXPENSES INCLUDING BUT NOT LIMITED
 * TO ANY INCIDENTAL, SPECIAL, INDIRECT, PUNITIVE OR CONSEQUENTIAL DAMAGES, LOST
 * PROFITS OR LOST DATA, COST OF PROCUREMENT OF SUBSTITUTE GOODS, TECHNOLOGY,
 * SERVICES, OR ANY CLAIMS BY THIRD PARTIES (INCLUDING BUT NOT LIMITED TO ANY
 * DEFENSE THEREOF), OR OTHER SIMILAR COSTS.
 *
 * Should you have any questions regarding your right to use this Software,
 * contact MbientLab via email: hello@mbientlab.com
 */

#import "LogViewController.h"
#import "LogEntry.h"
#import "DeviceConfiguration.h"

#import "JBLineChartView.h"
#import "JBBarChartView.h"
#import "JBChartHeaderView.h"
#import "JBChartInformationView.h"

#import "MBProgressHUD.h"

@interface LogViewController () <JBBarChartViewDataSource, JBBarChartViewDelegate, DeviceDelegate>
@property (weak, nonatomic) IBOutlet JBChartHeaderView *headerView;
@property (weak, nonatomic) IBOutlet JBBarChartView *barChartView;
@property (weak, nonatomic) IBOutlet JBChartInformationView *informationView;
@property (weak, nonatomic) IBOutlet UIProgressView *progressBar;
@property (weak, nonatomic) IBOutlet UILabel *statusLabel;
@property (weak, nonatomic) IBOutlet UISwitch *demoSwitch;

@property (nonatomic, strong) MBLMetaWear *device;
@property (nonatomic, strong) NSString *logFilename;
@property (nonatomic, strong) NSMutableArray *logData;
@property (nonatomic, strong) NSMutableArray *chartData;
@property (nonatomic, strong) NSMutableArray *timeData;
@property (nonatomic) BOOL doingReset;
@end

// Bar chart size and position constants
static CGFloat const BarChartViewControllerBarPadding = 1.0f;
static NSInteger const BarChartViewControllerNumBars = 60;

// Used for Demo Mode data generation
static NSInteger const BarChartViewControllerMaxBarHeight = 100;
static NSInteger const BarChartViewControllerMinBarHeight = 0;

// Estimate of steps per mile assuming casual walking speed @150 pounds
static NSInteger const StepsPerMile = 2000;

@implementation LogViewController
@synthesize logFilename = _logFilename;

- (NSString *)logFilename
{
    if (!_logFilename) {
        NSArray *paths = NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES);
        if (paths.count) {
            _logFilename = [NSString stringWithFormat:@"%@/archive_logfile", paths[0]];
        } else {
            [[[UIAlertView alloc] initWithTitle:@"Error" message:@"Cannot find documents directory, logging not supported" delegate:nil cancelButtonTitle:@"Okay" otherButtonTitles:nil] show];
            _logFilename = nil;
        }
    }
    return _logFilename;
}

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    //Set up navigation bar colors
    self.navigationController.navigationBar.tintColor = ColorNavigationTint;
    self.navigationController.navigationBar.barTintColor = ColorNavigationBarTint;
    self.navigationController.navigationBar.translucent = TRUE;
    [self.navigationController.navigationBar setTitleTextAttributes:[NSDictionary dictionaryWithObjectsAndKeys:ColorNavigationTitle,NSForegroundColorAttributeName, FontNavigationTitle, NSFontAttributeName, nil]];
    self.navigationItem.backBarButtonItem = [[UIBarButtonItem alloc] initWithTitle:@"" style:UIBarButtonItemStylePlain target:nil action:nil];
   
    // Creating bar chart and defining its properties
    self.barChartView.delegate = self;
    self.barChartView.dataSource = self;
    self.barChartView.minimumValue = 0.0f;
    self.barChartView.inverted = NO;
    self.barChartView.backgroundColor = ColorBarChartBackground;
    
    // Setup title and header
    self.headerView.titleLabel.text = @"Steps in Last Hour";
    self.headerView.separatorColor = ColorBarChartHeaderSeparatorColor;
    
    // Load saved values and display
    self.logData = [[NSKeyedUnarchiver unarchiveObjectWithFile:self.logFilename] mutableCopy];
    if (!self.logData) {
        self.logData = [NSMutableArray arrayWithCapacity:BarChartViewControllerNumBars];
    }
    [self.barChartView reloadData];
}

- (void)viewWillAppear:(BOOL)animated
{
    [super viewWillAppear:animated];
    
    [self updateInterface];
}

- (IBAction)connectionButtonPressed:(id)sender
{
    if ([self.navigationItem.leftBarButtonItem.title isEqualToString:@"Connect"]) {
        [self performSegueWithIdentifier:@"ShowScanScreen" sender:nil];
    } else {
        MBLMetaWear *device = self.device;
        self.device = nil;
        
        [device connectWithHandler:^(NSError *error) {
            [device setConfiguration:nil handler:nil];
        }];
        [device forgetDevice];
        
        [self clearLogPressed:nil];
        [self updateInterface];
    }
}

- (void)updateInterface
{
    [[[MBLMetaWearManager sharedManager] retrieveSavedMetaWearsAsync] success:^(NSArray<MBLMetaWear *> * _Nonnull array) {
        if (array.count) {
            self.device = array[0];
            self.statusLabel.text = @"Logging...";
            [self.navigationItem.leftBarButtonItem setTitle:@"Remove"];
            [self refreshPressed:self];
        } else {
            [self.navigationItem.leftBarButtonItem setTitle:@"Connect"];
            self.statusLabel.text = @"No MetaWear Connected";
        }
    }];
}

- (void)updateHeader
{
    // Set calorie label to new value
    self.headerView.subtitleLabel.text = [NSString stringWithFormat:@"Active Calories burned: %d", [self totalCaloriesBurned]];
}

- (int)totalCaloriesBurned
{
    float totalCalories = 0;
    for (LogEntry *entry in self.logData) {
        totalCalories += entry.calories;
    }
    return totalCalories;
}

- (IBAction)refreshPressed:(id)sender
{
    if (!self.device) {
        return;
    }
    
    if (self.demoSwitch.on) {
        return;
    }
    
    self.statusLabel.text = @"Connecting...";
    [self.device connectWithHandler:^(NSError *error) {
        if (error) {
            [[[UIAlertView alloc] initWithTitle:@"Error" message:@"Cannot connect to logger, make sure it is charged and within range" delegate:nil cancelButtonTitle:@"Okay" otherButtonTitles:nil] show];
            return;
        }
        
        self.progressBar.hidden = NO;
        self.progressBar.progress = 0;
        self.statusLabel.text = @"Syncing...";
        
        DeviceConfiguration *configuration = self.device.configuration;
        configuration.delegate = self;
        [configuration startDownload];
    }];
}

- (void)downloadDidUpdateProgress:(float)number
{
    [self.progressBar setProgress:number animated:YES];
}

- (void)downloadDidRecieveEntry:(LogEntry *)entry
{
    [self.logData addObject:entry];
    if (self.logData.count > BarChartViewControllerNumBars) {
        [self.logData removeObjectAtIndex:0];
    }
}

- (void)downloadCompleteWithError:(NSError *)error
{
    if (error) {
        [[[UIAlertView alloc] initWithTitle:@"Error" message:error.localizedDescription delegate:nil cancelButtonTitle:@"Okay" otherButtonTitles:nil] show];
    }
    [NSKeyedArchiver archiveRootObject:self.logData toFile:self.logFilename];
    
    if (!self.doingReset) {
        [self updateHeader];
        [self.barChartView reloadData];
    } else {
        self.doingReset = NO;
        [self clearLogPressed:nil];
    }
    self.progressBar.hidden = YES;
    self.statusLabel.text = @"Logging...";
    // We have our data so get outta here
    [self.device disconnectWithHandler:nil];
}

- (IBAction)switchChanged:(id)sender
{
    if (self.demoSwitch.on) {
        // Create some random data
        self.logData = [NSMutableArray arrayWithCapacity:BarChartViewControllerNumBars];
        for (int i = 0; i < BarChartViewControllerNumBars; i++) {
            NSNumber *rand = [NSNumber numberWithFloat:MAX((BarChartViewControllerMinBarHeight), arc4random() % (BarChartViewControllerMaxBarHeight)) * 1000];
            NSDate *timestamp = [NSDate dateWithTimeIntervalSinceNow:(BarChartViewControllerNumBars - i) * -60];
            [self.logData addObject:[[LogEntry alloc] initWithTotalRMS:rand timestamp:timestamp]];
        }
    } else {
        // Reload the real data
        self.logData = [[NSKeyedUnarchiver unarchiveObjectWithFile:self.logFilename] mutableCopy];
    }
    [self updateHeader];
    [self.barChartView reloadData];
}

- (IBAction)clearLogPressed:(id)sender
{
    [[NSFileManager defaultManager] removeItemAtPath:self.logFilename error:nil];
    [self.logData removeAllObjects];
    [self updateHeader];
    [self.barChartView reloadData];
}

- (IBAction)resetDevicePressed:(id)sender
{
    if (!self.device) {
        return;
    }

    MBProgressHUD *hud = [MBProgressHUD showHUDAddedTo:self.view animated:YES];
    MBLMetaWear *device = self.device;
    [device connectWithHandler:^(NSError *error) {
        if (error) {
            [[[UIAlertView alloc] initWithTitle:@"Error" message:error.localizedDescription delegate:nil cancelButtonTitle:@"Okay" otherButtonTitles:nil] show];
            return;
        }
        [device setConfiguration:device.configuration handler:^(NSError *error) {
            if (error) {
                [[[UIAlertView alloc] initWithTitle:@"Error" message:error.localizedDescription delegate:nil cancelButtonTitle:@"Okay" otherButtonTitles:nil] show];
                return;
            }
            [device disconnectWithHandler:^(NSError *error) {
                if (error) {
                    [[[UIAlertView alloc] initWithTitle:@"Error" message:error.localizedDescription delegate:nil cancelButtonTitle:@"Okay" otherButtonTitles:nil] show];
                    return;
                }
                [hud hide:YES];
            }];
        }];
    }];
}


- (NSUInteger)numberOfBarsInBarChartView:(JBBarChartView *)barChartView
{
    // If there are not enough samples, then add zeros to the array until there are enough samples and give a warning message.
    if (self.logData.count < BarChartViewControllerNumBars) {
        //[[[UIAlertView alloc] initWithTitle:@"Error" message:@"Not enough samples, chart data will be limited." delegate:nil cancelButtonTitle:@"Okay" otherButtonTitles:nil] show];
        int emptyEntries = BarChartViewControllerNumBars - (int)self.logData.count;
        for (int i = 0; i < emptyEntries; i++){
            [self.logData insertObject:[[LogEntry alloc] init] atIndex:0];
        }
    }
    return BarChartViewControllerNumBars;
}

- (void)barChartView:(JBBarChartView *)barChartView didSelectBarAtIndex:(NSUInteger)index touchPoint:(CGPoint)touchPoint
{
    LogEntry *entry = self.logData[index];
    [self.informationView setTitleText:entry.titleText];
    [self.informationView setValueText:entry.valueText unitText:nil];
    [self.informationView setHidden:NO animated:YES];
}

- (void)didDeselectBarChartView:(JBBarChartView *)barChartView
{
    [self.informationView setHidden:YES animated:YES];
}

#pragma mark - JBBarChartViewDelegate

- (CGFloat)barChartView:(JBBarChartView *)barChartView heightForBarViewAtIndex:(NSUInteger)index
{
    LogEntry *entry = self.logData[index];
    return entry.steps;
}

- (UIColor *)barChartView:(JBBarChartView *)barChartView colorForBarViewAtIndex:(NSUInteger)index
{
    return (index % 2 == 0) ? ColorBarChartBarOrange : ColorBarChartBarRed;
}

- (UIColor *)barSelectionColorForBarChartView:(JBBarChartView *)barChartView
{
    return [UIColor whiteColor];
}

- (CGFloat)barPaddingForBarChartView:(JBBarChartView *)barChartView
{
    return BarChartViewControllerBarPadding;
}

@end

File: ActivityTracker/LogViewController.swift
//
//  LogViewController.swift
//  ActivityTracker
//
//  Created by Stephen Schiffli on 4/25/17.
//  Copyright © 2017 MBIENTLAB Inc. All rights reserved.
//

import UIKit
import MetaWear
import MBProgressHUD

// Bar chart size and position constants
fileprivate let BarChartViewControllerBarPadding: CGFloat = 1.0
fileprivate let BarChartViewControllerNumBars = 60

// Used for Demo Mode data generation
fileprivate let BarChartViewControllerMaxBarHeight = 100
fileprivate let BarChartViewControllerMinBarHeight = 0

// Estimate of steps per mile assuming casual walking speed @150 pounds
fileprivate let StepsPerMile = 2000

class LogViewController: JBBaseChartViewController {
    @IBOutlet weak var headerView: JBChartHeaderView!
    @IBOutlet weak var barChartView: JBBarChartView!
    @IBOutlet weak var informationView: JBChartInformationView!
    @IBOutlet weak var progressBar: UIProgressView!
    @IBOutlet weak var statusLabel: UILabel!
    @IBOutlet weak var demoSwitch: UISwitch!
    
    var device: MBLMetaWear!
    let logFilename: String = {
        let paths = NSSearchPathForDirectoriesInDomains(.documentDirectory, .userDomainMask, true)
        if let path = paths.first {
            return "\(path)/archive_logfile"
        }
        print("Cannot find documents directory, logging not supported")
        return ""
    }()
    var logData: [LogEntry]!
    var chartData: Int!
    var timeData: Int!
    var totalCaloriesBurned: Int {
        return Int(logData.reduce(0.0) { (accum, entry) in
            return accum + entry.calories
        })
    }
    var doingReset: Bool = false
    
    
    override func viewDidLoad() {
        super.viewDidLoad()

        // Do any additional setup after loading the view.
        // Set up navigation bar colors
        navigationController?.navigationBar.tintColor = ColorNavigationTint
        navigationController?.navigationBar.barTintColor = ColorNavigationBarTint
        navigationController?.navigationBar.isTranslucent = true
        navigationController?.navigationBar.titleTextAttributes = [
            NSForegroundColorAttributeName: ColorNavigationTitle,
            NSFontAttributeName: FontNavigationTitle
        ]
        navigationItem.backBarButtonItem = UIBarButtonItem(title: "", style: .plain, target: nil, action: nil)

        // Creating bar chart and defining its properties
        barChartView.delegate = self
        barChartView.dataSource = self
        barChartView.minimumValue = 0.0
        barChartView.isInverted = false
        barChartView.backgroundColor = ColorBarChartBackground

        // Setup title and header
        headerView.titleLabel.text = "Steps in Last Hour"
        headerView.separatorColor = ColorBarChartHeaderSeparatorColor

        // Load saved values and display
        logData = NSKeyedUnarchiver.unarchiveObject(withFile: logFilename) as? [LogEntry] ?? []
        barChartView.reloadData()
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        
        updateInterface()
    }
    
    func updateInterface() {
        MBLMetaWearManager.shared().retrieveSavedMetaWearsAsync().success { (array) in
            if array.count > 0 {
                self.device = array.firstObject as! MBLMetaWear
                self.statusLabel.text = "Logging..."
                self.navigationItem.leftBarButtonItem?.title = "Remove"
                //refreshPressed(self)
            } else {
                self.navigationItem.leftBarButtonItem?.title = "Connect"
                self.statusLabel.text = "No MetaWear Connected"
            }
        }
    }
    
    func updateHeader() {
        headerView.subtitleLabel.text = "Active Calories burned: \(totalCaloriesBurned)"
    }
    
    @IBAction func connectionButtonPressed(_ sender: Any) {
        if navigationItem.leftBarButtonItem?.title == "Connect" {
           performSegue(withIdentifier: "ShowScanScreen", sender: nil)
        } else {
            device.connectAsync().success { (device) in
                device.setConfigurationAsync(nil)
            }
            device.forgetDevice()
            device = nil
            
            //clearLogPressed(nil)
            updateInterface()
        }
    }

    @IBAction func refreshPressed(_ sender: Any) {
        guard let device = device else {
            return
        }
        guard !demoSwitch.isOn else {
            return
        }
        statusLabel.text = "Connecting..."
        device.connectAsync().success { (device) in
            self.progressBar.isHidden = false
            self.progressBar.progress = 0.0
            self.statusLabel.text = "Syncing..."
            
            let configuration = self.device.configuration as! DeviceConfiguration
            configuration.delegate = self
            configuration.startDownload()
        }.failure { (error) in
            let alert = UIAlertController(title: "Error", message: "Cannot connect to logger, make sure it is charged and within range", preferredStyle: .alert)
            alert.addAction(UIAlertAction(title: "Okay", style: .default))
            self.present(alert, animated: true)
        }
    }
    
    @IBAction func switchChanged(_ sender: Any) {
        if demoSwitch.isOn {
            // Create some random data
            logData = []
            for i in 0..<BarChartViewControllerNumBars {
                let randOffset = Int(arc4random()) % (BarChartViewControllerMaxBarHeight - BarChartViewControllerMinBarHeight)
                let timestamp = Date(timeIntervalSinceNow: TimeInterval((BarChartViewControllerNumBars - i) * -60))
                logData.append(LogEntry(steps: BarChartViewControllerMinBarHeight + randOffset, timestamp: timestamp))
            }
        } else {
            // Reload the real data
            logData = NSKeyedUnarchiver.unarchiveObject(withFile: logFilename) as? [LogEntry] ?? []
        }
        updateHeader()
        barChartView.reloadData()
    }
    
    @IBAction func clearLogPressed(_ sender: Any) {
        try? FileManager.default.removeItem(atPath: logFilename)
        logData.removeAll()
        updateHeader()
        barChartView.reloadData()
    }
    
    @IBAction func resetDevicePressed(_ sender: Any) {
        guard let device = device else {
            return
        }
        let hud = MBProgressHUD.showAdded(to: view, animated: true)
        device.connectAsync().continueOnDispatchWithSuccess(block: { (t) -> Any? in
            return device.setConfigurationAsync(device.configuration)
        }).continueOnDispatchWithSuccess(block: { (t) -> Any? in
            return device.disconnectAsync()
        }).continueOnDispatch { (t) -> Any? in
            hud.hide(animated: true)
            if let error = t.error {
                let alert = UIAlertController(title: "Error", message: error.localizedDescription, preferredStyle: .alert)
                alert.addAction(UIAlertAction(title: "Okay", style: .default))
                self.present(alert, animated: true)
            }
            return nil
        }
    }
}

extension LogViewController: DeviceDelegate {
    func downloadDidUpdateProgress(_ number: Float) {
        progressBar.setProgress(number, animated: true)
    }
    
    func downloadDidRecieveEntry(_ entry: LogEntry) {
        logData.append(entry)
        if logData.count > BarChartViewControllerNumBars {
            logData.remove(at: 0)
        }
    }
    
    func downloadCompleteWithError(_ error: Error?) {
        if let error = error {
            let alert = UIAlertController(title: "Error", message: error.localizedDescription, preferredStyle: .alert)
            alert.addAction(UIAlertAction(title: "Okay", style: .default))
            present(alert, animated: true)
        }
        NSKeyedArchiver.archiveRootObject(logData, toFile: logFilename)
        if !doingReset {
            updateHeader()
            barChartView.reloadData()
        } else {
            doingReset = false
            //clearLogPressed(nil)
        }
        progressBar.isHidden = true
        statusLabel.text = "Logging..."
        // We have our data so get outta here
        device.disconnectAsync()
    }
}

extension LogViewController: JBBarChartViewDataSource {
    func numberOfBars(in barChartView: JBBarChartView!) -> UInt {
        if logData.count < BarChartViewControllerNumBars {
            let emptyEntries = BarChartViewControllerNumBars - logData.count
            for _ in 0..<emptyEntries {
                logData.insert(LogEntry(), at: 0)
            }
        }
        return UInt(BarChartViewControllerNumBars)
    }
    
    func barChartView(_ barChartView: JBBarChartView!, didSelectBarAt index: UInt) {
        let entry = logData[Int(index)]
        informationView.setTitleText(entry.titleText)
        informationView.setValueText(entry.valueText, unitText: nil)
        informationView.setHidden(false, animated: true)
    }
    
    func didDeselect(_ barChartView: JBBarChartView!) {
        informationView.setHidden(true, animated: true)
    }
}

extension LogViewController: JBBarChartViewDelegate {
    func barChartView(_ barChartView: JBBarChartView!, heightForBarViewAt index: UInt) -> CGFloat {
        return CGFloat(logData[Int(index)].steps)
    }

    func barChartView(_ barChartView: JBBarChartView!, colorForBarViewAt index: UInt) -> UIColor! {
        return index % 2 == 0 ? ColorBarChartBarOrange : ColorBarChartBarRed
    }
    func barSelectionColor(for barChartView: JBBarChartView!) -> UIColor! {
        return UIColor.white
    }
    func barPadding(for barChartView: JBBarChartView!) -> CGFloat {
        return BarChartViewControllerBarPadding
    }
}
